File: backend/app/models/voice_profile.py
# backend/app/models/voice_profile.py

from enum import Enum
from typing import Optional
from pydantic import BaseModel


class ToneStyle(str, Enum):
    FORMAL = "formal"
    PROFESSIONAL = "professional"
    FRIENDLY = "friendly"
    CASUAL = "casual"
    STARTUP_CASUAL = "startup_casual"


class AddressStyle(str, Enum):
    DIRECT_YOU = "direct_you"
    THIRD_PERSON = "third_person"
    WE_LOOKING = "we_looking"


class SentenceStyle(str, Enum):
    SHORT_PUNCHY = "short_punchy"
    BALANCED = "balanced"
    DETAILED = "detailed"


class StructurePreferences(BaseModel):
    """Preferences for JD structure and section ordering."""
    lead_with_benefits: bool = False
    section_order: list[str] = ["intro", "responsibilities", "requirements", "benefits"]
    include_salary_prominently: bool = False


class VoiceProfile(BaseModel):
    """Enhanced voice profile with Voice DNA attributes."""
    id: str
    name: str

    # Tone (enhanced from single enum to spectrum + description)
    tone_formality: int = 3  # 1=very formal, 5=very casual
    tone_description: str = "Professional"  # AI-extracted or user-written

    # Legacy fields (kept for compatibility)
    tone: ToneStyle = ToneStyle.PROFESSIONAL
    address_style: AddressStyle = AddressStyle.DIRECT_YOU
    sentence_style: SentenceStyle = SentenceStyle.BALANCED

    # Structure (new)
    structure_preferences: StructurePreferences = StructurePreferences()
    structure_preference: str = "mixed"  # Legacy field

    # Vocabulary
    words_to_avoid: list[str] = []
    words_to_prefer: list[str] = []

    # Brand personality (new)
    brand_values: list[str] = []  # e.g., ["innovation", "collaboration"]

    # Source tracking (new)
    source_examples: list[str] = []  # JDs used to create this profile
    created_via: str = "manual"  # "examples" | "guided" | "manual"

    # Metadata
    example_jd: Optional[str] = None  # Legacy single example
    is_default: bool = False

    def to_prompt_context(self) -> str:
        """Generate prompt context for AI from this profile."""
        formality_descriptions = {
            1: "Very formal and corporate",
            2: "Professional and polished",
            3: "Professional but approachable",
            4: "Friendly and conversational",
            5: "Casual and energetic",
        }

        address_descriptions = {
            AddressStyle.DIRECT_YOU: 'Address candidates directly using "you" and "your".',
            AddressStyle.THIRD_PERSON: 'Use third person like "the candidate" or "the ideal person".',
            AddressStyle.WE_LOOKING: 'Frame from company perspective: "We\'re looking for someone who..."',
        }

        sentence_descriptions = {
            SentenceStyle.SHORT_PUNCHY: "Use short, punchy sentences. Maximum 2-3 sentences per paragraph.",
            SentenceStyle.BALANCED: "Use balanced sentence length. Mix of short and medium sentences.",
            SentenceStyle.DETAILED: "Use detailed, thorough sentences. Complete explanations.",
        }

        parts = [
            f"VOICE PROFILE: {self.name}",
            f"Tone: {self.tone_description} ({formality_descriptions.get(self.tone_formality, 'Professional')})",
            f"Address Style: {address_descriptions[self.address_style]}",
            f"Sentence Style: {sentence_descriptions[self.sentence_style]}",
        ]

        # Structure preferences
        if self.structure_preferences.lead_with_benefits:
            parts.append("Structure: Lead with benefits and impact before listing requirements.")

        if self.structure_preferences.section_order:
            parts.append(f"Section Order: {' → '.join(self.structure_preferences.section_order)}")

        if self.structure_preferences.include_salary_prominently:
            parts.append("Include salary/compensation information prominently near the top.")

        # Vocabulary
        if self.words_to_avoid:
            parts.append(f"Words to AVOID: {', '.join(self.words_to_avoid)}")

        if self.words_to_prefer:
            parts.append(f"Words to PREFER: {', '.join(self.words_to_prefer)}")

        # Brand values
        if self.brand_values:
            parts.append(f"Brand Values to Reflect: {', '.join(self.brand_values)}")

        # Legacy example
        if self.example_jd:
            parts.append(f"Example JD for reference style:\n{self.example_jd[:1000]}")

        return "\n".join(parts)


class VoiceProfileCreate(BaseModel):
    """Input model for creating a voice profile."""
    name: str
    tone_formality: int = 3
    tone_description: str = "Professional"
    tone: ToneStyle = ToneStyle.PROFESSIONAL
    address_style: AddressStyle = AddressStyle.DIRECT_YOU
    sentence_style: SentenceStyle = SentenceStyle.BALANCED
    structure_preferences: StructurePreferences = StructurePreferences()
    structure_preference: str = "mixed"
    words_to_avoid: list[str] = []
    words_to_prefer: list[str] = []
    brand_values: list[str] = []
    source_examples: list[str] = []
    created_via: str = "manual"
    example_jd: Optional[str] = None
    is_default: bool = False


class ExtractVoiceRequest(BaseModel):
    """Input for extracting voice profile from example JDs."""
    example_jds: list[str]
    suggested_name: Optional[str] = None

File: frontend/src/types/voice-profile.ts
// frontend/src/types/voice-profile.ts

export type ToneStyle = 'formal' | 'professional' | 'friendly' | 'casual' | 'startup_casual';
export type AddressStyle = 'direct_you' | 'third_person' | 'we_looking';
export type SentenceStyle = 'short_punchy' | 'balanced' | 'detailed';
export type CreationMethod = 'examples' | 'guided' | 'manual';
export type RuleType = 'exclude' | 'include' | 'format' | 'order' | 'limit' | 'custom';

export interface ProfileRule {
  id: string;
  text: string; // Natural language rule: "Never include salary"
  ruleType: RuleType;
  target?: string; // What it applies to: "salary", "requirements"
  value?: string; // Additional value: "bullets", "5"
  source: 'manual' | 'ai_suggested' | 'extracted';
  active: boolean;
}

export interface SuggestedRule {
  text: string;
  ruleType: RuleType;
  target?: string;
  value?: string;
  confidence: number;
  evidence: string;
}

export interface StructurePreferences {
  leadWithBenefits: boolean;
  sectionOrder: string[];
  includeSalaryProminently: boolean;
}

export interface VoiceProfile {
  id: string;
  name: string;

  // Tone (enhanced)
  toneFormality: number; // 1-5
  toneDescription: string;
  tone: ToneStyle; // legacy

  // Style
  addressStyle: AddressStyle;
  sentenceStyle: SentenceStyle;

  // Structure (new)
  structurePreferences: StructurePreferences;
  structurePreference: string; // legacy

  // Vocabulary
  wordsToAvoid: string[];
  wordsToPrefer: string[];

  // Brand (new)
  brandValues: string[];

  // Source tracking (new)
  sourceExamples: string[];
  createdVia: CreationMethod;

  // Dynamic rules (new)
  rules: ProfileRule[];
  formatGuidance?: string;

  // Metadata
  exampleJd?: string; // legacy
  isDefault: boolean;
  createdAt?: string;
}

export interface VoiceProfileFormData {
  name: string;
  toneFormality: number;
  toneDescription: string;
  tone: ToneStyle;
  addressStyle: AddressStyle;
  sentenceStyle: SentenceStyle;
  structurePreferences: StructurePreferences;
  wordsToAvoid: string;
  wordsToPrefer: string;
  brandValues: string;
  structurePreference: string;
}

// Voice extraction response (from API)
export interface VoiceExtractionResult {
  tone: ToneStyle;
  toneFormality: number;
  toneDescription: string;
  addressStyle: AddressStyle;
  sentenceStyle: SentenceStyle;
  structureAnalysis: {
    leadsWithBenefits: boolean;
    typicalSectionOrder: string[];
    includesSalary: boolean;
  };
  vocabulary: {
    commonlyUsed: string[];
    notablyAvoided: string[];
  };
  brandSignals: {
    values: string[];
    personality: string;
  };
  suggestedRules: SuggestedRule[];
  formatGuidance?: string;
  summary: string;
}

// UI Options
export const TONE_OPTIONS = [
  { value: 'formal', label: 'Formal & Traditional' },
  { value: 'professional', label: 'Professional' },
  { value: 'friendly', label: 'Friendly & Warm' },
  { value: 'casual', label: 'Casual' },
  { value: 'startup_casual', label: 'Startup Casual' },
] as const;

export const ADDRESS_OPTIONS = [
  { value: 'direct_you', label: 'Direct (You/Your)' },
  { value: 'third_person', label: 'Third Person (The candidate)' },
  { value: 'we_looking', label: "Company Voice (We're looking for)" },
] as const;

export const SENTENCE_OPTIONS = [
  { value: 'short_punchy', label: 'Short & Punchy' },
  { value: 'balanced', label: 'Balanced' },
  { value: 'detailed', label: 'Detailed & Thorough' },
] as const;

export const STRUCTURE_OPTIONS = [
  { value: 'bullet_heavy', label: 'Bullet-Heavy' },
  { value: 'mixed', label: 'Mixed' },
  { value: 'paragraph_focused', label: 'Paragraph-Focused' },
] as const;

export const FORMALITY_LABELS: Record<number, string> = {
  1: 'Very Formal',
  2: 'Professional',
  3: 'Balanced',
  4: 'Friendly',
  5: 'Casual',
};

/**
 * Maps rule targets and keywords to form field names.
 * Used for validation to check if profile rules reference missing data.
 */
export const RULE_TARGET_FIELD_MAP: Record<string, { field: string; label: string }> = {
  // Salary-related
  salary: { field: 'salaryRange', label: 'Salary Range' },
  compensation: { field: 'salaryRange', label: 'Salary Range' },
  pay: { field: 'salaryRange', label: 'Salary Range' },
  // Benefits-related
  benefits: { field: 'benefits', label: 'Benefits' },
  perks: { field: 'benefits', label: 'Benefits' },
  // Team/company-related
  team: { field: 'teamSize', label: 'Team Size' },
  team_size: { field: 'teamSize', label: 'Team Size' },
  // Location-related
  location: { field: 'location', label: 'Location' },
  remote: { field: 'location', label: 'Location' },
  office: { field: 'location', label: 'Location' },
  // Company culture
  culture: { field: 'companyDescription', label: 'Company Description' },
  company: { field: 'companyDescription', label: 'Company Description' },
  mission: { field: 'companyDescription', label: 'Company Description' },
};

/**
 * Create default structure preferences
 */
export function createDefaultStructurePreferences(): StructurePreferences {
  return {
    leadWithBenefits: false,
    sectionOrder: ['intro', 'responsibilities', 'requirements', 'benefits'],
    includeSalaryProminently: false,
  };
}

/**
 * Create a default voice profile (for migration/initialization)
 */
export function createDefaultVoiceProfile(partial: Partial<VoiceProfile> = {}): VoiceProfile {
  const defaults: VoiceProfile = {
    id: '',
    name: '',
    toneFormality: 3,
    toneDescription: 'Professional',
    tone: 'professional',
    addressStyle: 'direct_you',
    sentenceStyle: 'balanced',
    structurePreferences: createDefaultStructurePreferences(),
    structurePreference: 'mixed',
    wordsToAvoid: [],
    wordsToPrefer: [],
    brandValues: [],
    sourceExamples: [],
    createdVia: 'manual',
    rules: [],
    isDefault: false,
  };

  return {
    ...defaults,
    ...partial,
    // Ensure arrays are never undefined after spread
    rules: partial.rules ?? defaults.rules,
  };
}
